pygfx/utils/transform.py

```diff
@@ -0,0 +1,449 @@
+
+        if matrix is None:
+            matrix = np.eye(4, dtype=float)
+
+        self.untracked_matrix = np.asarray(matrix)
+
+        if position is not None:
+            self.position = position
+
+        if rotation is not None:
+            self.rotation = rotation
+
+        if scale is not None:
+            self.scale = scale
+
+    def flag_update(self):
+        self.last_modified = perf_counter_ns()
+        super().flag_update()
+
+    @property
+    def matrix(self):
+        view_array = self.untracked_matrix.view()
+        view_array.flags.writeable = False
+        return view_array
+
+    @matrix.setter
+    def matrix(self, value):
+        self.untracked_matrix[:] = value
+        self.flag_update()
+
+    def __matmul__(self, other):
+        if isinstance(other, AffineTransform):
+            return AffineTransform(self.matrix @ other.matrix)
+
+        return np.asarray(self) @ other
+
+    def __array__(self, dtype=None):
+        return self.untracked_matrix.astype(dtype, copy=False)
+
+
+class ChainedTransform(AffineBase):
+    def __init__(
+        self, transform_sequence: List[AffineTransform], *, settable_index: int = None
+    ) -> None:
+        super().__init__()
+        self.sequence = transform_sequence
+        self.settable: AffineTransform = None
+        self.before: AffineBase = None
+        self.after: AffineBase = None
+
+        if settable_index is not None:
+            idx = settable_index
+            before, after = self.sequence[:idx], self.sequence[idx:]
+            target, after = after[0], after[1:]
+
+            self.settable = target
+
+            if len(before) > 0:
+                self.before = ChainedTransform(before)
+            else:
+                self.before = AffineTransform()
+
+            if len(after) > 0:
+                self.after = ChainedTransform(after)
+            else:
+                self.after = AffineTransform()
+
+    @property
+    def last_modified(self):
+        return max(x.last_modified for x in self.sequence)
+
+    @cached
+    def _matrix(self):
+        return la.matrix_combine(self.sequence)
+
+    @property
+    def matrix(self):
+        return self._matrix
+
+    @matrix.setter
+    def matrix(self, value):
+        if self.settable is None:
+            raise AttributeError(
+                "This ChainedTransform doesn't use `settable_index` and "
+                "thus can't set properties."
+            )
+
+        new_value = self.before.inverse_matrix @ value @ self.after.inverse_matrix
+        self.settable.matrix = new_value
+
+    def __matmul__(self, other):
+        if isinstance(other, ChainedTransform):
+            return ChainedTransform(self.sequence + other.sequence)
+        elif isinstance(other, AffineTransform):
+            return ChainedTransform(self.sequence + [other])
+        else:
+            return np.asarray(self) @ other
+
+
+class ImplicitTransform(AffineBase):
```

Comment:
This also seems to be unused now, so it can be removed?

Reply:
Same as `ChainedTransform`. It's an artifact of an earlier version of the transform system and no longer in use. Happy to remove it, too.